### PR TITLE
[QDP] Add angle and basis encoding tests for numpy input

### LINKED ISSUE #1178: Implement Angle & Basis encoding test
Implement the missing test cases at [https://github.com/apache/mahout/blob/main/testing/qdp/test_numpy.py#L139](https://www.google.com/search?q=https://github.com/apache/mahout/blob/main/testing/qdp/test_numpy.py%23L139)

### PR DESCRIPTION
**Pull request by yehfelareborn**
### Changes

- [ ] Bug fix
- [ ] New feature
- [ ] Refactoring
- [ ] Documentation
- [x] Test
- [ ] CI/CD pipeline
- [ ] Other

### Why

- angle and basis encoding were already implemented in Rust (angle.rs, basis.rs), but had no Python test coverage
- The existing test_encode_numpy_encoding_methods only covered amplitude and left a TODO comment for the missing methods

### How

Add angle encoding tests: 1D single sample, 2D multi-sample, loading from .npy file, and a negative test for wrong input size
Add basis encoding tests: 1D single sample, 2D multi-sample, loading from .npy file
Add basis state correctness test: verify the encoded output is the correct one-hot vector (|basis_index⟩)
Add basis negative tests: negative index, non-integer index, out-of-range index

Remove the TODO comment

## Checklist

- [ ] Added or updated unit tests for all changes
- [ ] Added or updated documentation for all changes

### CODE REVIEW

File: testing/qdp/test_numpy.py [hunk 1]
```diff
@@ -147,6 +146,255 @@ def test_encode_numpy_encoding_methods(encoding_method):
...
+@requires_qdp
+@pytest.mark.gpu
+@pytest.mark.parametrize("num_qubits", [1, 2, 3, 4])
+def test_encode_numpy_angle_encoding_1d(num_qubits):
+    """Angle encoding: 1D array with num_qubits angles (single sample)"""
+    from _qdp import QdpEngine
+
+    pytest.importorskip("torch")
+    if not torch.cuda.is_available():
+        pytest.skip("GPU required for QdpEngine")
```
[the remark this answers is not in the document]
Comment (ryankert01) [reply testing/qdp/test_numpy.py]: +1

File: testing/qdp/test_numpy.py [hunk 2]
```diff
@@ -147,6 +146,255 @@ def test_encode_numpy_encoding_methods(encoding_method):
...
+@requires_qdp
+@pytest.mark.gpu
+@pytest.mark.parametrize(
+    ("bad_data", "description"),
+    [
+        (np.array([[-1.0]], dtype=np.float64), "negative index"),
+        (np.array([[0.5]], dtype=np.float64), "non-integer index"),
+    ],
+)
+def test_encode_numpy_basis_encoding_invalid_index(bad_data, description):
```
Comment (rich7420) [inline testing/qdp/test_numpy.py]: Nit: The description parameter is injected by parametrize but never used in the test body. Consider removing it from the signature and using pytest.param(..., id=...) instead for test ID clarity.

File: testing/qdp/test_numpy.py [hunk 3]
```diff
@@ -147,6 +146,255 @@ def test_encode_numpy_encoding_methods(encoding_method):
...
+        pytest.skip("GPU required for QdpEngine")
+
+    engine = QdpEngine(device_id=0)
+    # Angle encoding expects exactly num_qubits values: one angle per qubit
+    angles = np.random.uniform(0, 2 * np.pi, size=num_qubits).astype(np.float64)
+
+    qtensor = engine.encode(angles, num_qubits, encoding_method="angle")
+    tensor = torch.from_dlpack(qtensor)
+
+    _verify_tensor(tensor, (1, 2**num_qubits), check_normalization=True)
```
Comment (rich7420) [inline testing/qdp/test_numpy.py:159]: nit: The angle encoding tests verify normalization but don't check the actual amplitude values. Consider adding a correctness test similar to test_encode_numpy_basis_state_correctness — for example, with 1 qubit and a known angle θ, verify that the output probabilities match [cos²(θ/2), sin²(θ/2)].

**Review by rich7420 — COMMENTED** [review]
@yehfelareborn thanks for the patch!
left some comments below

File: testing/qdp/test_numpy.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yehfelareborn) [reply testing/qdp/test_numpy.py]: I resolve this issue.thanks

File: testing/qdp/test_numpy.py (hunk 2 above)
Comment (yehfelareborn) [reply testing/qdp/test_numpy.py]: Hi, i removed the description part and replaced it with pytest.param, thanks

File: testing/qdp/test_numpy.py (hunk 3 above)
Comment (yehfelareborn) [reply testing/qdp/test_numpy.py:159]: I added 3 cases — θ = 0, π/2, and π — to verify that the output probabilities match [cos²(θ/2), sin²(θ/2)]. thanks

**Review by 400Ping — CHANGES_REQUESTED** [review]
Thanks for adding the NumPy coverage here. The overall direction makes sense, but I think there is one blocking correctness issue and one docs/API consistency issue.

**Angle encoding correctness**
The new 1-qubit correctness test currently expects probabilities `[cos²(theta/2), sin²(theta/2)]`.

However, the current QDP implementation and docs both describe angle encoding as a product state of the form:

  cos(x_k)|0> + sin(x_k)|1>

So for 1 qubit, the expected probabilities should be `[cos²(theta), sin²(theta)]`, not half-angle probabilities.

Concretely:
- `qdp/qdp-kernels/src/angle.cu` multiplies amplitudes by `cos(angle)` / `sin(angle)`
- `docs/qdp/concepts.md` documents the same semantics

**Comment by yehfelareborn** [comment]
Thanks for pointing out the mistake I made. The fix has already been committed for full-angle semantics: cos(θ)|0⟩ + sin(θ)|1⟩

**Review by 400Ping — APPROVED** [review]
Overall LGTM

**Comment by 400Ping** [comment]
tag me when CI passes

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -12,7 +12,7 @@ repos:
 
   # Python linter and formatter
   - repo: https://github.com/astral-sh/ruff-pre-commit
-    rev: v0.15.5
+    rev: v0.15.6
```
Comment (ryankert01) [inline .pre-commit-config.yaml]: question of why introducing this change? Do you encounter something?

Comment (yehfelareborn) [reply .pre-commit-config.yaml]: I may have encountered a pre-commit issue. Either way, I'm withdrawing this edit.

**Comment by yehfelareborn** [comment]
> tag me when CI passes

@400Ping CI is passed
